jellyfin/cordova-plugin-chromecast
# 1.0.0 Refactor (Issue #36)

I am excited to present to you the 1.0.0 re-write of the plugin!  The plugin now follows the chrome.cast api much more closely!  And (as far as I am aware) works without issue!

I say re-write because I estimated that ~80% of Android was changed.  And iOS was moved to objc to fit the cordova plugin standard (so a ~100% rewrite).

There should be very few changes required to migrate from 0.x to 1.0.0.  Mostly if you depended on `getRouteListElement` you will have to switch to `scanForRoutes`/`stopScan`, or if you depended on behavior not found on chrome desktop's api implementation.

These changes have been tested on:
- Android 4.4, 7.x
- iOS 9.3 (iPad 3), 10.3 (iPhone 5), 12.4.2 (iPad mini 2), 13.2.2 (emulator iPad 7, emulator iPhone 8)
- Chrome desktop browser 78.0 (the tests pass on this version of chrome, which demonstrates that the same code will work on all 3 platforms [basically])

The largest changes of this refactor are:

- [x] iOS, Android, and Chrome Desktop all work the same now!  (With a few caveats.)
All 3 will pass the same set of tests included in the plugin.

- [x] iOS and Android have `scanForRoutes`, `stopScan`, and `selectRoute` to replace `getRouteListElement`

- [x] iOS and Android still do not implement the full chromecast API, but a few more have been added, and existing ones have been made to match chrome desktop behavior.

**Bug Fixes**: 
- Fixes #17 (cancel should return error)
- Fixes #22 for Android (getRouteListElement() returns empty object) non-issue now
- Fixes #22 for iOS (getRouteListElement() returns empty object) non-issue now
- Fixes #23 Metadata not set
- Fixes #26 #27 (compile/dependancy error)
- Fixes #32 (duplicate routes are displayed)
- Fixes #36 (Because that is this issue :p)
- Fixes #43 (crash on requestSession)
- Fixes #41 remove deprecated `sendJavascript` usage
- Fixes #47 iOS crash
- Fixes #48 selectRoute crash
- Fixes #49 selectRoute silent fail
- Fixes #50 Switch to Mocha for tests
- Play audio files https://github.com/miloproductionsinc/cordova-plugin-chromecast/issues/3

**Improvements:**
- Upgraded to CAF v3, this got rid of a huge amount of deprecation notices.
- Added eslint tests to ensure uniform js code stlying
- Added a style checker for java
- Updated documentation on: how to run style checkers, how to run tests.
- Added a pull request template which instructions prospective pull requesters to run eslint and the tests
- Repaired and added to the existing tests enforcing chrome desktop SDK behavior.
- On app initialize you will receive any previously started+active session.  (This includes navigating to a new page, or force killing the app and re-loading.)
- Implemented startRouteScan and stopRouteScan to replace getRoutes ([see why](https://github.com/jellyfin/cordova-plugin-chromecast/issues/22#issuecomment-530773677))
- fixed selectRoute for Android 4.4.2
- Allow native error responses to send a json object "code" and "description" parameters in addition to just the error code string. 
- Implement basic queue functionality
- Tests can run on chrome desktop browser as well (with a couple stubs to enable missing functionality)
  - This is super useful because you can write tests and check that they work correctly on chrome first.  Which is very important when chrome's implementation is what we are implementing.
- Added a basic code example
- Updated readme with all kinds of instructions (differences between chrome, api list, plugin-specific api list, api usage example, support+quirks, setup, how to test, how to run formatting test)
- Added pull request template
- Added scripts to assist in enforcing code formatting for JS and Java

**Auto-tests**:
- initialize
    - receiverListener
    - sessionListener
- leaveSession
- stopSession
- loadMedia Video
- media.setVolume
- media.play
- media.pause
- media.stop
- session.setReceiverVolumeLevel
- session.setReceiverMuted
- requestSession
- cordova.startRouteScan
- cordova.stopRouteScan
- cordova.selectRoute
- MediaMetadata (All types with most of their special params)
- loadMedia Audio (Did not test playback controls)
- loadMedia Image
- load queue
- jump to queue item

**Manual Tests**:
- Start session, reload page, call initialize, do we get session again?
- Start session, force kill app, restart app, do we get session again?
- Start session, join session with additional device, **leave** session on original, session still continues?, force kill and restart original device, original device should not rejoin the session automatically.
- for session Leave: probably need to join the session from an additional device,then leave the session, restart app, does it rejoin the session? (it shouldn't) (pre-req rejoin on restart must work)
- for loadmedia from external sender: start session on device, join session on other device, loadmedia from other device, is session.addMediaListener triggered?
- for external exit: Start session, kill session with other device, requestSession should show the the join dialog, not stopCasting

In addition to all the issues listed above, this probably fixes these as well:
- #52 
- #53
